### PR TITLE
Implement Tag and Untag features

### LINKED ISSUE #85: As a user, I can add tags to modules
so that I can label and organise my modules more effectively.

### LINKED ISSUE #86: As a user, I can add tags to lectures
so that I can label and organise my lectures more effectively.

### LINKED ISSUE #87: As a user, I can tag videos
so that I can label and organise my videos more effectively.

### PR DESCRIPTION
**Pull request by lennoxtr**
close #85
close #86 
close #87

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/ParseArgument.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+     * @throws ParseException if the given {@code name} is invalid.
+     */
+    public static ModuleCode parseModule(String moduleCode) throws ParseException {
+        requireNonNull(moduleCode);
+        String trimmedModuleCode = moduleCode.trim();
+        if (!ModuleCode.isValidCode(trimmedModuleCode)) {
+            throw new ParseException(ModuleCode.MESSAGE_CONSTRAINTS);
+        }
+        return new ModuleCode(moduleCode);
+    }
```
Comment (hingen) [inline src/main/java/seedu/address/logic/parser/ParseArgument.java:36]: Seems to be duplicated code, could you use `ParseUtil.parseModuleCode(String)`?

As much as possible, don't create dummy `Module` objects just for the sake of retrieving the real `Module` object. `Model` has a method for retrieving `Module` using `ModuleCode`.

File: src/main/java/seedu/address/logic/parser/ParseArgument.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+     */
+
+    public static LectureName parseLecture(String lectureName) throws ParseException {
+        requireNonNull(lectureName);
+        String trimmedLectureName = lectureName.trim();
+        if (!LectureName.isValidName(lectureName)) {
+            throw new ParseException(LectureName.MESSAGE_CONSTRAINTS);
+        }
+        return new LectureName(trimmedLectureName);
+    }
```
Comment (hingen) [inline src/main/java/seedu/address/logic/parser/ParseArgument.java:52]: Same as before. Could you use `Parseutil.parseLectureName(String)`?

File: src/main/java/seedu/address/logic/parser/ParseArgument.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+     */
+
+    public static VideoName parseVideo(String videoName) throws ParseException {
+        requireNonNull(videoName);
+        String trimmedVideoName = videoName.trim();
+        if (!VideoName.isValidName(trimmedVideoName)) {
+            throw new ParseException(VideoName.MESSAGE_CONSTRAINTS);
+        }
+        return new VideoName(trimmedVideoName);
+    }
```
Comment (hingen) [inline src/main/java/seedu/address/logic/parser/ParseArgument.java:68]: Same as before, consider using `ParseUtil.parseVideoName(String)`.

File: src/main/java/seedu/address/logic/parser/ParseArgument.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+                .map(trimmedtag -> new Tag(trimmedtag))
+                .collect(Collectors.toList());
+
+        for (Tag tag : listOfTags) {
+            if (!Tag.isValidTagName(tag.tagName)) {
+                throw new ParseException(Tag.MESSAGE_CONSTRAINTS);
+            }
+        }
+        return new HashSet<>(listOfTags);
+    }
```
Comment (hingen) [inline src/main/java/seedu/address/logic/parser/ParseArgument.java:107]: Consider moving these to `seedu.address.logic.parser.ParserUtil.java`

File: src/main/java/seedu/address/logic/parser/ParseArgument.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+import seedu.address.model.lecture.LectureName;
+import seedu.address.model.module.ModuleCode;
+import seedu.address.model.tag.Tag;
+import seedu.address.model.video.VideoName;
+
+
+/**
+ * A dummy class containing utility methods used for parsing strings into ModuleCode, LectureName, and Video.
+ */
+public class ParseArgument {
```
Comment (hingen) [inline src/main/java/seedu/address/logic/parser/ParseArgument.java:21]: This class seems to serve the same purpose as `seedu.address.logic.parser.ParserUtil.java`

File: src/main/java/seedu/address/model/Model.java [hunk 6]
```diff
@@ -213,13 +223,13 @@ public interface Model {
...
      *
-     * @param lecture The lectre to check if t contains the video.
-     * @param videoName The video name to check if exist.
-     * @return True if a video with the name {@code videoName} exists in {@code lecture}. Otherwise, false.
+     * @param lecture The lecture to check if it contains the video.
+     * @param video The video to check if exist.
+     * @return True if a video with the same name as {@code video} exists in {@code lecture}. Otherwise, false.
      */
-    boolean hasVideo(ReadOnlyLecture lecture, VideoName videoName);
+    boolean hasVideo(ReadOnlyLecture lecture, VideoName video);
```
Comment (hingen) [inline src/main/java/seedu/address/model/Model.java:232]: Don't think this parameter name should be changed.

File: src/main/java/seedu/address/model/Model.java [hunk 7]
```diff
@@ -260,11 +271,7 @@ public interface Model {
      */
     void setVideo(ReadOnlyLecture lecture, Video target, Video editedVideo);
 
-    /**
-     * Returns an unmodifiable view of the filtered module list.
-     *
-     * @return An unmodifiable view of the filtered module list.
-     */
+
```
Comment (hingen) [inline src/main/java/seedu/address/model/Model.java:274]: Probably should leave this as is, don't delete the javadocs.

File: src/main/java/seedu/address/model/Model.java [hunk 8]
```diff
@@ -232,6 +242,7 @@ public interface Model {
      */
     Video getVideo(ReadOnlyLecture lecture, VideoName videoName);
 
+
```
Comment (hingen) [inline src/main/java/seedu/address/model/Model.java:245]: Please delete this newline.

File: src/main/java/seedu/address/model/Model.java [hunk 9]
```diff
@@ -165,6 +165,15 @@ public interface Model {
...
 
+    /**
+     * Returns true if a {@code Module} has a lecture with name {@code lecture}.
+     *
+     * @param module The module to check if it contains the lecture.
+     * @param lecture The lecture to check if exist.
+     * @return True if a lecture with the same name as {@code lecture} exists in {@code module}. Otherwise, false.
+     */
+    boolean hasLecture(ReadOnlyModule module, LectureName lecture);
+
```
Comment (hingen) [inline src/main/java/seedu/address/model/Model.java:176]: Don't think this is necessary. Consider using the `hasLecture(ModuleCode, LectureName)` method.

File: src/main/java/seedu/address/model/ModelManager.java
Comment (hingen) [inline src/main/java/seedu/address/model/ModelManager.java:1]: Might want to consider undoing these changes. May have been an artifact from resolving a merge conflict due to very outdated code.

File: src/main/java/seedu/address/model/Model.java
Comment (hingen) [inline src/main/java/seedu/address/model/Model.java:1]: Might want to consider undoing these changes. May have been an artifact from resolving a merge conflict due to very outdated code.

File: src/main/java/seedu/address/model/module/Module.java [hunk 10]
```diff
@@ -130,7 +131,7 @@ public void addLecture(Lecture lecture) {
      * @throws DuplicateLectureException Indicates that {@code editedLecture} is the same as another existing
      *                                   lecture in the module.
      */
-    public void setLecture(ReadOnlyLecture target, Lecture editedLecture) {
+    public void setLecture(ReadOnlyLecture target, Lecture editedLecture) throws LectureNotFoundException {
```
Comment (hingen) [inline src/main/java/seedu/address/model/module/Module.java:134]: No need for this as `LectureNotFoundException` is a runtime exception. But yes, thanks for importing the `LectureNotFoundException`, was a bug that I missed  this out.

**Comment by lennoxtr** [comment]
close as there's an updated PR

**Pull request closed without merging**